runtime/common/src/check_nonce.rs

```diff
@@ -0,0 +1,285 @@
+// This file is part of Acala.
+
+// Copyright (C) 2020-2021 Acala Foundation.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use codec::{Decode, Encode};
+use frame_support::weights::DispatchInfo;
+use module_support::AddressMapping;
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{DispatchInfoOf, Dispatchable, One, SignedExtension},
+	transaction_validity::{
+		InvalidTransaction, TransactionLongevity, TransactionValidity, TransactionValidityError, ValidTransaction,
+	},
+};
+use sp_std::vec;
+
+/// Nonce check and increment to give replay protection for transactions.
+///
+/// # Transaction Validity
+///
+/// This extension affects `requires` and `provides` tags of validity, but DOES NOT
+/// set the `priority` field. Make sure that AT LEAST one of the signed extension sets
+/// some kind of priority upon validating transactions.
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct CheckNonce<T: frame_system::Config + module_evm::Config>(
+	#[codec(compact)] pub T::Index,
+	#[codec(skip)] pub bool, /* should check evm nonce */
+);
```

Comment:
I think enum is better.
```
pub enum CheckNonce<T: frame_system::Config + module_evm::Config>(
	Nonce(#[codec(compact)] pub T::Index),
	EvmNonce(#[codec(skip)] pub T::Index),
);
```
Only change this:
https://github.com/AcalaNetwork/Acala/blob/78dfe2bb82/ts-tests/tests/test-sign-eth.ts#L122

Reply:
enum will break compatibility with existing extrinsic format. could have a custom Encode/Decode implementation to fix that

Reply:
I think we should just do named fields